src/pip/_internal/utils/misc.py

```diff
@@ -880,15 +880,29 @@ def split_auth_from_netloc(netloc):
     return netloc, user_pass
 
 
+def redact_password_from_url(url):
+    def _redact_netloc(netloc):
+        netloc, (user, passw) = split_auth_from_netloc(netloc)
+        if not user:
+            return netloc
+        passw = '' if passw is None else ':****'
+        return '{user}{passw}@{netloc}'.format(user=user,
+                                               passw=passw,
+                                               netloc=netloc)
+
+    return transform_url(url, _redact_netloc)
+
+
 def remove_auth_from_url(url):
     # Return a copy of url with 'username:password@' removed.
     # username/pass params are passed to subversion through flags
     # and are not recognized in the url.
+    return transform_url(url, lambda netloc: split_auth_from_netloc(netloc)[0])
 
-    # parsed url
-    purl = urllib_parse.urlsplit(url)
-    netloc, user_pass = split_auth_from_netloc(purl.netloc)
 
+def transform_url(url, transform_netlock):
```

Comment:
I would move this function to be above `redact_password_from_url()` and `remove_auth_from_url()` (and after `redact_netloc()`), since it is used in both of them, and it is nice to have a progression so that functions within a module depend only on functions that precede them (so it can be read from top to bottom).